src/Altinn.App.Core/Features/Correspondence/CorrespondenceClient.cs

```diff
@@ -0,0 +1,241 @@
+                requestException?.HttpStatusCode.ToString() ?? "Unknown",
+                requestException?.ResponseBody ?? "No response body"
+            );
+
+            activity?.Errored(e, requestException?.ProblemDetails?.Detail);
+            throw;
+        }
+        catch (Exception e)
+        {
+            activity?.Errored(e);
+            _logger.LogError(e, "Failed to fetch correspondence status: {Exception}", e);
+            throw new CorrespondenceRequestException($"Failed to fetch correspondence status: {e}", e);
+        }
+    }
+
+    private async Task<HttpRequestMessage> AuthenticatedHttpRequestFactory(
+        HttpMethod method,
+        string uri,
+        HttpContent? content,
+        CorrespondencePayloadBase payload
+    )
+    {
+        _logger.LogDebug("Fetching access token via factory");
+        AccessToken accessToken = await AuthorisationFactory(payload);
+
+        _logger.LogDebug("Constructing authorized http request for target uri {TargetEndpoint}", uri);
+        HttpRequestMessage request = new(method, uri) { Content = content };
+
+        request.Headers.Authorization = new AuthenticationHeaderValue(TokenTypes.Bearer, accessToken);
+        request.Headers.TryAddWithoutValidation(General.SubscriptionKeyHeaderName, _platformSettings.SubscriptionKey);
+
+        return request;
+    }
+
+    private ProblemDetails? GetProblemDetails(string responseBody)
+    {
+        if (string.IsNullOrWhiteSpace(responseBody))
+        {
+            return null;
+        }
+
+        try
+        {
+            return JsonSerializer.Deserialize<ProblemDetails>(responseBody);
+        }
+        catch (Exception e)
+        {
+            _logger.LogError("Error parsing ProblemDetails from correspondence api: {Error}", e);
+        }
+
+        return null;
+    }
+
+    private string GetUri(string relativePath)
+    {
+        string baseUri = _platformSettings.ApiCorrespondenceEndpoint.TrimEnd('/');
+        return $"{baseUri}/{relativePath.TrimStart('/')}";
+    }
+
+    private async Task<TContent> HandleServerCommunication<TContent>(
+        HttpRequestMessage request,
+        CancellationToken cancellationToken
+    )
+    {
+        using HttpClient client = _httpClientFactory.CreateClient();
+        HttpResponseMessage response = await client.SendAsync(request, cancellationToken);
+        string responseBody = await response.Content.ReadAsStringAsync(cancellationToken);
+
+        if (response.StatusCode != HttpStatusCode.OK)
+        {
+            var problemDetails = GetProblemDetails(responseBody);
+            throw new CorrespondenceRequestException(
+                $"Correspondence request failed with status {response.StatusCode}: {problemDetails?.Detail}",
+                problemDetails,
+                response.StatusCode,
+                responseBody
+            );
+        }
+
+        _logger.LogDebug("Correspondence request succeeded: {Response}", responseBody);
```

Comment:
Use of debug logging should also be something we should discuss in the team, as in how often and how we do it (there are more perfomant APIs for this as well)

Reply:
I've got no useful opinions on this. What's here now is what made sense to me during testing 🤷